sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java

```diff
@@ -622,20 +628,56 @@ public void setSessionState(String sessionId, byte[] sessionState) {
     @Override
     public void close() {
         asyncClient.close();
+
+        if (messageProcessor.get() != null) {
+            messageProcessor.get().dispose();
+        }
+
+        Disposable activeSubscription = messageProcessorSubscription.get();
+        if (activeSubscription != null) {
+            activeSubscription.dispose();
+        }
     }
 
     /**
-     * Given an {@code emitter}, queues that work in {@link SynchronousMessageSubscriber}. If the synchronous job has
-     * not been created, will initialise it.
+     * Given an {@code emitter}, creates a {@link EmitterProcessor} to receive messages from Service Bus. If the
+     * message processor has not been created, will initialise it.
      */
     private void queueWork(int maximumMessageCount, Duration maxWaitTime,
         FluxSink<ServiceBusReceivedMessageContext> emitter) {
-        final long id = idGenerator.getAndIncrement();
-        final SynchronousReceiveWork work = new SynchronousReceiveWork(id, maximumMessageCount, maxWaitTime,
-            emitter);
-        final SynchronousMessageSubscriber syncSubscriber = new SynchronousMessageSubscriber(work);
 
-        logger.info("[{}]: Started synchronous message subscriber.", id);
-        asyncClient.receive(DEFAULT_RECEIVE_OPTIONS).subscribeWith(syncSubscriber);
+        if (messageProcessor.get() != null && messageProcessor.get().isDisposed()) {
+            logger.error("[{}]: Can not receive messaged because client is closed.", asyncClient.getEntityPath());
+            return;
+        }
+
+        if (messageProcessor.get() == null) {
+            EmitterProcessor<ServiceBusReceivedMessageContext> processor = asyncClient.receive(DEFAULT_RECEIVE_OPTIONS)
+                .subscribeWith(EmitterProcessor.create(false));
+
+            if (!messageProcessor.compareAndSet(null, processor)) {
+                processor.dispose();
+            }
+
+            logger.info("[{}]: Started ContinuesMessageSubscriber message subscriber for entity.",
+                asyncClient.getEntityPath());
+        }
+
+        Disposable newSubscription = messageProcessor.get()
+            .take(maximumMessageCount)
+            .timeout(maxWaitTime)
+            .doOnNext(messageContext -> {
```

Comment:
These are for side effects.

Reply:
doOnNext, doOnError, doOnComplete are for side effects like logging or monitoring. It shouldn't be used as part of the main chain of operations.